[WIP] Implement iNatSounds models

tom2208 opened the pull request:
### Models
iNatSounds used three models which we want to implement if not implemented already:
* [x] ViT-B-16 (WiP)
* [x] ResNet50

### Preprocessing
* [x] 22.050 Hz sample rate
* [x] Hann window size 
  * [x] 512 size / 256 ?
  * [x]  stride length of 128 / 32 ?
  * [ ] 1024 FFT length for the STFT
* [x] mel-scale 
  * [x] maps frequences in the range [50 Hz, 11.015 Hz] (paper says 11.015 kHz i think that is a mistake)
  * [x] 128 log spaced mel bins or 298?
  * [x] convert resulting magnitudes to decibels with a [0, 255] span representing a gray scal image
* [x] fixed window size of ~3 sec (512 pixel)
  * [x] for longer audio process sec windows that are strided by 1.5 sec (256 pixel)
  * [ x] for smaller recordings pad with silence (zeroes) to 3 seconds
* [x] for standardization all spectograms are resized from 512x128x3 RGB to 224x224x3
  * [x] to make use of the pretrained backbone the gray sacale image is duplicated to get an rgb one
* [x] fix input length 
* [x] edit chunk generating in a case that input is exactly 3 sec
### check preprocessing
* [x] check if all the parameters are same as the author script
* [ ] still need to compare with iNatsound paper by different runs
   * [x] linear probing on pow (it stops after 4 epoch  (https://wandb.ai/deepbirddetect/BioFoundation/runs/vit_inatsound_linearprobing_BirdSet_POW_1_2025-06-13_121753/workspace?nw=nwuserpariavalizadeh)
   * [x] finetuning on pow (https://wandb.ai/deepbirddetect/BioFoundation/groups/biofoundation_inatsound_finetune_BirdSet__POW/overview)
   * [ ] ablations for pooling

### Links
* the paper: [https://openreview.net/pdf?id=QCY01LvyKm](https://openreview.net/pdf?id=QCY01LvyKm)
* pretrained models download: https://drive.google.com/drive/folders/1u8iqzP2WL2nkTMp9VZ5FCa3zveOrPj2X
* the code they used for training: [https://github.com/cvl-umass/iNatSounds](https://github.com/cvl-umass/iNatSounds)
  * Geo model is defined here: https://github.com/cvl-umass/iNatSounds/blob/main/src/models.py

tom2208 committed:
reformated files with black AND added initial vit.py model

tom2208 committed:
some vit things

tom2208 committed:
self.model.model -> self.model; .-.

tom2208 committed:
resolved some errors

tom2208 committed:
resolved some errors encountered new ones

tom2208 committed:
tryng to implement preprocessing for vit has errors though

tom2208 commented:
Current error is a shape error:

Shape bofore preprocessing: `torch.Size([32, 1, 110250])`
Shape after preprocessing:`torch.Size([32, 1, 858, 128])`

Important part:
```shell
AssertionError: Wrong image height! Expected 224 but got 858!
```

```shell
Error executing job with overrides: ['logger=wandb_biofoundation', 'experiment=birdset/fewshot/vit.yaml']
Traceback (most recent call last):
  File "/workspace/birdset/train.py", line 110, in train
    trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.get("ckpt_path"))
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 543, in fit
    call._call_and_handle_interrupt(
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 579, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 986, in _run
    results = self._run_stage()
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 1028, in _run_stage
    self._run_sanity_check()
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 1057, in _run_sanity_check
    val_loop.run()
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/loops/utilities.py", line 182, in _decorator
    return loop_run(self, *args, **kwargs)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/loops/evaluation_loop.py", line 135, in run
    self._evaluation_step(batch, batch_idx, dataloader_idx, dataloader_iter)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/loops/evaluation_loop.py", line 396, in _evaluation_step
    output = call._call_strategy_hook(trainer, hook_name, *step_args)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/trainer/call.py", line 311, in _call_strategy_hook
    output = fn(*args, **kwargs)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/lightning/pytorch/strategies/strategy.py", line 411, in validation_step
    return self.lightning_module.validation_step(*args, **kwargs)
  File "/workspace/birdset/modules/base_module.py", line 214, in validation_step
    val_loss, preds, targets = self.model_step(batch, batch_idx)
  File "/workspace/birdset/modules/base_module.py", line 176, in model_step
    logits = self.forward(**batch)
  File "/workspace/birdset/modules/base_module.py", line 145, in forward
    return self.model.forward(*args, **kwargs)
  File "/workspace/projects/biofoundation/modules/models/vit.py", line 79, in forward
    return self.model(spectograms)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/torchvision/models/vision_transformer.py", line 291, in forward
    x = self._process_input(x)
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/torchvision/models/vision_transformer.py", line 271, in _process_input
    torch._assert(h == self.image_size, f"Wrong image height! Expected {self.image_size} but got {h}!")
  File "/home/vscode/.cache/pypoetry/virtualenvs/birdset-xS3fZVNL-py3.10/lib/python3.10/site-packages/torch/__init__.py", line 1209, in _assert
    assert condition, message
AssertionError: Wrong image height! Expected 224 but got 858!
```

tom2208 commented:
I found the problem, i implemented the whole process wrong, in the wrong order. I need to first create a 512 x 128 x 3 image using mel etc and then resize it. I resized it beforehand and then applied mel.

However i get 110250 samples with a sample rate of 22050 = 5 sec of audio. The paper wants 3 seconds. In the `/module/network: vit.yaml` i defined `sampling_rate=22050` and `length=3` (3 seconds i reckon?).

I think this is no problem for implementing since i need to resize it anyway but will influence the results badly.

tom2208 committed:
end of the day, does not work yet

raphaelschwinger commented:
@tom2208 Check if the `max_length` property of the `BaseTransforms` class is correctly set. The idea is that this value gets set by the `module.network.length` config parameter.

raphaelschwinger commented:
@tom2208 For augmentations: use the ones from the BirdSet pipeline, since we want to evaluate the pretrained model it's not necessary to use the exact same augmentations.

tom2208 committed:
vit for inat implemented

The-Mats committed:
Added vit finetuning config

The-Mats committed:
Additions to vit

tom2208 commented:
Current results for ViT are looking weird: https://wandb.ai/deepbirddetect/BioFoundation/runs/vit_finetune_BirdSet_HSN_1_2025-04-17_123753?nw=nwuserstu216509

Validation cmAP is about 0.5, in tests it just gets below 0.05.

raphaelschwinger commented:
@tom2208 The validation cmAP should be around 90% and the loss should go down < 0.02. I guess that the preprocessing is not done correctly and the model therefore struggles to learn something meaningful. Debug this. In this process you could also implement the preprocessing using `torchaudio`. Take a look at https://github.com/DBD-research-group/BirdSet/blob/main/notebooks/tutorials/model_inference.ipynb

tom2208 committed:
added resnet

The-Mats committed:
Merge branch 'feature/inatsounds' of github.com:DBD-research-group/BioFoundation into feature/inatsounds

The-Mats committed:
Vit fixes, new preprocessing (WIP)

The-Mats committed:
Merge branch 'main' into feature/inatsounds

The-Mats committed:
Finally Merge branch 'main' into feature/inatsounds

The-Mats committed:
Small fix for table script

The-Mats committed:
Fixed the fix :)

The-Mats committed:
Even better fix and new table

The-Mats committed:
Small fix and updated birdset table

The-Mats committed:
Small vit changes

tom2208 committed:
trying something

tom2208 committed:
first try implementing the preprocessing from inat

PariaValizadeh committed:
Merge remote-tracking branch 'origin/main' into feature/inatsounds

PariaValizadeh committed:
same vit as main

PariaValizadeh committed:
vit for inatsoundmodel

tom2208 committed:
starting to reproduce inat code they provided

PariaValizadeh committed:
Merge remote-tracking branch 'origin/main' into feature/inatsounds

PariaValizadeh committed:
update birdset to biofoundationmodule, add length for audio input in nocall_sampler

PariaValizadeh committed:
update chunk

PariaValizadeh committed:
change vit to inatsound in finetuning

PariaValizadeh committed:
Merge remote-tracking branch 'origin/main' into feature/inatsounds

PariaValizadeh committed:
update names

PariaValizadeh committed:
Merge remote-tracking branch 'origin/main' into feature/inatsounds

PariaValizadeh committed:
add pooling

PariaValizadeh committed:
update experiences configs

PariaValizadeh committed:
Merge remote-tracking branch 'origin/main' into feature/inatsounds

PariaValizadeh committed:
add pooling test config

PariaValizadeh committed:
update tags

raphaelschwinger commented on projects/biofoundation/configs/experiment/birdset/linearprobing/pooling/pooling_config.yaml:
```diff
@@ -0,0 +1,104 @@
+# @package _global_
+#package global is neccessary!
```
You could have used the one from `birdset/linearprobing/validation/validation_config.yaml` is there a difference?

PariaValizadeh replied:
the only difference is tags and nocall sampler in data module

raphaelschwinger commented on projects/biofoundation/modules/models/vit_inatsound.py:
```diff
@@ -0,0 +1,273 @@
+            for name, param in self.model.named_parameters():
+                if "heads.head" not in name:
+                    param.requires_grad = False
+
```
This should be implemented in the parent classes. Remove the duplication.

raphaelschwinger commented on projects/biofoundation/modules/models/vit_inatsound.py:
```diff
@@ -0,0 +1,273 @@
+    def _load_model(self) -> nn.Module:
+        model = models.vit_b_16(weights=None)
+
+        checkpoint_path = "/workspace/models/vit/vit_single_mixup.pt"
```
use the `checkpoint_path` init parameter so that it is possible to load another checkpoint from the config.

raphaelschwinger commented on projects/biofoundation/modules/models/vit_inatsound.py:
```diff
@@ -0,0 +1,273 @@
+            if orig_sr != TARGET_SR:
+                resampler = T.Resample(orig_sr, TARGET_SR)
+                waveform = resampler(waveform)
+            return waveform
```
remove the resampling as it is done in the datamodule

raphaelschwinger commented on projects/biofoundation/modules/models/vit_inatsound.py:
```diff
@@ -0,0 +1,273 @@
+                    end = start + WINDOW_SIZE
+                    chunk = pad_or_trim(waveform[..., start:end], WINDOW_SIZE)
+                    chunks.append(chunk)
+            return chunks
```
is also already done in the data pipeline

raphaelschwinger reviewed (approved):
@PariaValizadeh I looked through the code, looks good on the first glance! But I think you added some stuff that is not needed for our usecase.

tom2208 committed:
attempt on approximating inat preprocessing

tom2208 committed:
merging: refactoring preprocessing to vit_inatsound.py

tom2208 committed:
deleted code part that i missed earlier

tom2208 committed:
removed code i forgot to delete and the resampling part in preprocessing

tom2208 committed:
now just one proper preprocess function

PariaValizadeh committed:
add vit

PariaValizadeh committed:
Merge remote-tracking branch 'origin/feature/inatsounds' into feature/inatsounds

PariaValizadeh committed:
update preprocessing and model loading function

tom2208 committed:
T.mel and preprocess_old adapted to inat paper

tom2208 committed:
adjusted preprocess to the inat paper

tom2208 committed:
delet code i forgot to delete

tom2208 commented:
I adjusted the `preprocess` function to fit the inat paper. One author kindly answered my email: The "strange" values I was talking about which did not match the paper were just default values for a class, when they initialize the class they override those values with ones that match the paper.

I also adjusted some values. Like `power=1` since the inat paper wants magnitudes and not power (`power=2`). I will try to run experiments now.

tom2208 commented:
I tried an experiment with it:
./projects/biofoundation/train.sh experiment=birdset/linearprobing/pooling/vit_inatsound.yaml module.network.model.pooling=just_cls https://wandb.ai/deepbirddetect/BioFoundation/runs/vit_inatsound_just_cls_BirdSet_POW_1_2025-06-19_170940?nw=nwuserstu216509

It does not seem to learn much at all. If the preprocessing is at fault it may have something to do with the db conversion - at least that was the issue last time. The values in the db tensor seem oddly high. When i print max and min values of the not normalized values i get: -127 to +50 db which looks like an error?

raphaelschwinger commented:
closed in favour of #41
